### PR TITLE
Add build-time check to enforce PGO compilation for applicable DLL files

### LINKED ISSUE #12863: Introduce a PGO Verification Test
At a minimum, this test should verify that PGO is enabled for the builds that we expected it to be enabled for.  For example, on Windows, we can run "dumpbin /headers <coreclr.dll | clrjit.dll> | grep coffgrp" and validate that the build is a PGU build.

It's fine for developers to not build using PGO, but we do want to make sure that official and PR builds pass this test - otherwise, we need to figure out what happened.

NOTE: I have not yet found a way to detect whether or not PGO data was properly applied when using LLVM on Linux.  It appears that "llvm-profdata show" only operates on profdata files and not shared objects that have been built with profile data.

cc: @jorive

### PR DESCRIPTION
**Pull request by nategraf**
This PR includes an optional check, enabled with a new "enforcepgo" flag, after the compilation of the CLR to confirm that PGO was used, and fail the build if not. This is to catch release build configurations which break PGO

This is an alternative to PR #13127
Closes #12863

### CODE REVIEW

**Comment by nategraf** [comment]
@dotnet/rap-team

File: build.cmd [hunk 1]
```diff
@@ -193,6 +195,21 @@ if %__BuildTypeRelease%==1  set __BuildType=Release
...
+if %__EnforcePgo%==1 (
+    if %__BuildArchArm%==1 (
+        echo Error: enforcepgo cannot be used with arm architecture
+        goto Usage
+    )
+    if %__BuildArchArm64%==1 (
+        echo Error: enforcepgo cannot be used with arm64 architecture
+        goto Usage
+    )
+    if NOT %__BuildOS% == Windows_NT (
```
Comment (brianrob) [inline build.cmd]: I don't think you need to check the OS because build.cmd can only be run on Windows.

File: src/build.proj [hunk 2]
```diff
@@ -38,4 +38,19 @@
         DestinationFolder="$(BinDir)PDB" />
   </Target>
 
+  <PropertyGroup>
+    <RunEnforcePGO Condition="$(__EnforcePgo) == '1'">true</RunEnforcePGO>
+    <RunEnforcePGO Condition="$(__BuildOS) != 'Windows_NT'">false</RunEnforcePGO>
```
Comment (brianrob) [inline src/build.proj]: I don't think you need the OS check.

File: build.cmd (hunk 1 above)
Comment (nategraf) [inline build.cmd]: I added it with the thought that someone may build for Linux from Windows (like we used to do for tests). Is that not important for this case?

Comment (wtgodbe) [inline build.cmd]: Not for this case, __BuildOS will always be Windows_NT when running build.cmd

File: src/build.proj (hunk 2 above)
Comment (wtgodbe) [inline src/build.proj]: +1

**Comment by wtgodbe** [comment]
LGTM outside of Brian's comments, assuming you've tested the python script locally.

File: src/build.proj (hunk 2 above)
Comment (nategraf) [inline src/build.proj]: Removed in commit 7d4acb96db2533b60aa09b75ce783d34784a7165

File: build.cmd (hunk 1 above)
Comment (nategraf) [inline build.cmd]: Removed in commit 7d4acb96db2533b60aa09b75ce783d34784a7165

**Comment by nategraf** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test please

**Comment by nategraf** [comment]
@wtgodbe @brianrob Should I go ahead and merge this or do you have more comments / more suggested reviewers?

**Comment by brianrob** [comment]
@nategraf go ahead and merge.  Then you can enable this functionality in the CI and official builds in a subsequent PR.

**Pull request merged**